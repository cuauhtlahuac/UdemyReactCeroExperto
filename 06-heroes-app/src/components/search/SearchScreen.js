import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router-dom';

import HeroCard from 'components/heroes/HeroCard';
import { useForm } from 'hooks/useForm';
import { getHeroByName } from 'selectors/getHeroByName';

const SearchScreen = ({ history }) => {
	const { search } = useLocation();

	const { q: heroSearched = '' } = useMemo(() => queryString.parse(search), [
		search,
	]);

	const filteredHeroes = useMemo(() => getHeroByName(heroSearched), [
		heroSearched,
	]);

	const [ formValues, handleInputChange, reset ] = useForm({ searchText: '' });

	const { searchText } = formValues;

	const handleSearch = event => {
		event.preventDefault();
		history.push(`?q=${searchText}`);
		reset();
	};

	return (
		<div>
			<div className="row">
				<div className="col-5">
					<h3>Search</h3>
					<hr />

					<form onSubmit={handleSearch}>
						<input
							autoComplete="off"
							name="searchText"
							type="text"
							value={searchText}
							placeholder="Find your hero"
							className="form-control"
							onChange={handleInputChange}
						/>

						<button
							type="submit"
							className="btn my-2 btn-block btn-outline-primary"
						>
							Search
						</button>
					</form>
				</div>

				<div className="col-7">
					<h4>Results</h4>
					<hr />

					{heroSearched.length < 1 && (
						<div className="alert alert-info">
							You don't searching for heroes yet
						</div>
					)}

					{	( filteredHeroes.length < 1 &&
						heroSearched.length > 0 ) && 
						( <div className="alert alert-danger">
								Hero "{`${heroSearched}`}" don't found
							</div> )
					}

					{filteredHeroes.map(hero => <HeroCard key={hero.id} {...hero} />)}
				</div>
			</div>
		</div>
	);
};

export default SearchScreen;
