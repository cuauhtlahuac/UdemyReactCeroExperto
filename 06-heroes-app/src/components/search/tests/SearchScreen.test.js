import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
const { mount } = require('enzyme');

import SearchScreen from '../SearchScreen';

describe('<SearchScreen /> Tests', () => {
	const component = mount(
		<MemoryRouter initialEntries={[ '/search' ]}>
			<Route path="/" component={SearchScreen} />
		</MemoryRouter>,
	);

	test('should render correctly', () => {
		expect(component).toMatchSnapshot();
	});

	test('should show alert div', () => {
		expect(component.find('div.alert-info').text()).toBe(
			"You don't searching for heroes yet",
		);
	});

	describe('', () => {
		const component = mount(
			<MemoryRouter initialEntries={[ '/search?q=batman' ]}>
				<Route path="/search" component={SearchScreen} />
			</MemoryRouter>,
		);
		test('should update data when url change', () => {
			expect(component.find('div.alert-info').exists()).not.toBe(true);

			expect(component.find('h5.card-title').text()).toBe('Batman');
		});
		test('should clear the text input', () => {
			expect(component.find('input[name="searchText"]').prop('value')).toBe('');
		});
	});
});
