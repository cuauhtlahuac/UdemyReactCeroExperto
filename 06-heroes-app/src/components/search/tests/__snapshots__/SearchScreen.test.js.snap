// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchScreen /> Basic Tests should render correctly 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h3>
        Search
      </h3>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn my-2 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        You don't searching for heroes yet
      </div>
    </div>
  </div>
</div>
`;
